src/TesApi.Web/Events/RunnerEventsProcessor.cs

```diff
@@ -0,0 +1,516 @@
+                throw new ArgumentException("This message's Tags are missing.", nameof(message));
+            }
+
+            if (string.IsNullOrWhiteSpace(message.Event))
+            {
+                throw new ArgumentException("This message's event type is missing.", nameof(message));
+            }
+
+            if (message.Tags.Count == 0)
+            {
+                throw new ArgumentException("This message has no tags.", nameof(message));
+            }
+
+            if (message.Tags.ContainsKey(ProcessedTag))
+            {
+                throw new ArgumentException("This message was already processed.", nameof(message));
+            }
+
+            if (!message.Tags.ContainsKey("event-name") || !message.Tags.ContainsKey("task-id") || !message.Tags.ContainsKey("created"))
+            {
+                throw new ArgumentException("This message is missing required tags.", nameof(message));
+            }
+
+            // There are up to 10 tags allowed. We will be adding one.
+            // https://learn.microsoft.com/azure/storage/blobs/storage-manage-find-blobs?tabs=azure-portal#setting-blob-index-tags
+            if (message.Tags.Count > 9)
+            {
+                throw new ArgumentException("This message does not have space to add the processed tag.", nameof(message));
+            }
+        }
+
+        /// <summary>
+        /// Gets the details of this event message.
+        /// </summary>
+        /// <param name="message">Tes runner event message metadata.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> for controlling the lifetime of the asynchronous operation.</param>
+        /// <exception cref="DownloadOrParseException">Validation exceptions.</exception>
+        /// <returns>A <see cref="RunnerEventsMessage"/> containing the associated <seealso cref="Tes.Runner.Events.EventMessage"/>.</returns>
+        /// <remarks>This method assumes <paramref name="message"/> was successfully validated by <see cref="ValidateMessageMetadata(RunnerEventsMessage)"/>.</remarks>
+        public async Task<RunnerEventsMessage> DownloadAndValidateMessageContentAsync(RunnerEventsMessage message, CancellationToken cancellationToken)
+        {
+            Tes.Runner.Events.EventMessage content;
+
+            try
+            {
+                var messageText = await azureProxy.DownloadBlobAsync(message.BlobUri, cancellationToken);
+                content = System.Text.Json.JsonSerializer.Deserialize<Tes.Runner.Events.EventMessage>(messageText)
+                    ?? throw new DownloadOrParseException("Deserialize() returned null.");
+            }
+            catch (Exception ex)
+            {
+                throw new DownloadOrParseException($"Event message blob is malformed. {ex.GetType().FullName}:{ex.Message}", ex);
+            }
+
+            message = new(message, content);
+
+            // Validate content
+            Validate(Guid.TryParse(content.Id, out _),
```

Comment:
Because I was stress testing, I was looking for every possible place I could conceive things could break. Certainly, these two sections can be relaxed.